business/openid_auth.go

```diff
@@ -142,3 +326,142 @@ func GetOpenIdMetadata() (*OpenIdMetadata, error) {
+	requestParams.Set("code", openIdParams.Code)
+	requestParams.Set("grant_type", "authorization_code")
+	requestParams.Set("redirect_uri", redirect_uri)
+	response, err := httpClient.PostForm(openIdMetadata.TokenURL, requestParams)
+	if err != nil {
+		return fmt.Errorf("failure when requesting token from IdP: %w", err)
+	}
+
+	defer response.Body.Close()
+	rawTokenResponse, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return fmt.Errorf("failed to read token response from IdP: %w", err)
+	}
+
+	if response.StatusCode != 200 {
+		return fmt.Errorf("request failed (HTTP response status = %s)", response.Status)
+	}
+
+	// Parse token response
+	var tokenResponse struct {
+		IdToken string `json:"id_token"`
+	}
+
+	err = json.Unmarshal(rawTokenResponse, &tokenResponse)
+	if err != nil {
+		return fmt.Errorf("cannot parse OpenId token response: %w", err)
+	}
+
+	if len(tokenResponse.IdToken) == 0 {
+		return errors.New("the IdP did not provide an id_token")
+	}
+
+	openIdParams.IdToken = tokenResponse.IdToken
+	return nil
+}
+
+func ValidateOpenIdNonceCode(openIdParams *OpenIdCallbackParams) (validationFailure string) {
+	// Parse the received id_token from the IdP and check nonce code
+	idTokenClaims := openIdParams.ParsedIdToken.Claims.(jwt.MapClaims)
+	nonceHashHex := fmt.Sprintf("%x", openIdParams.NonceHash)
+	if nonceClaim, ok := idTokenClaims["nonce"]; !ok || nonceHashHex != nonceClaim.(string) {
+		validationFailure = "nonce code mismatch"
+	}
+	return
+}
+
+// - CSRF mitigation
+func ValidateOpenIdState(openIdParams *OpenIdCallbackParams) (validationFailure string) {
+	state := openIdParams.State
+
+	separator := strings.LastIndexByte(state, '-')
+	if separator != -1 {
+		csrfToken, timestamp := state[:separator], state[separator+1:]
+		csrfHash := sha256.Sum224([]byte(fmt.Sprintf("%s+%s+%s", openIdParams.Nonce, timestamp, config.GetSigningKey())))
+
+		if fmt.Sprintf("%x", csrfHash) != csrfToken {
+			//RespondWithError(w, http.StatusForbidden, "Request rejected because of CSRF mitigation.")
+			validationFailure = "CSRF mitigation"
+		}
+	} else {
+		validationFailure = "State parameter is invalid"
+	}
+
+	return
+}
+
+func VerifyOpenIdUserAccess(token string) (int, string, error) {
+	// Create business layer using the id_token
+	business, err := Get(token)
+	if err != nil {
+		return http.StatusInternalServerError, "Error instantiating the business layer", err
+	}
+
+	// Using the namespaces API to check if token is valid. In Kubernetes, the version API seems to allow
+	// anonymous access, so it's not feasible to use the version API for token verification.
+	nsList, err := business.Namespace.GetNamespaces()
+	if err != nil {
+		return http.StatusUnauthorized, "Token is not valid or is expired", err
+	}
+
+	// If namespace list is empty, return unauthorized error
+	if len(nsList) == 0 {
+		return http.StatusUnauthorized, "Not enough privileges to login", nil
```

Comment:
This has caused confusion with some users in the past. I think we should have a more explicit error message here. When we say "Not enough privileges to login" - what we specifically mean here is "User cannot get even a single namespace". Perhaps we should indicate this more explicitly (that they need to get able to get at least one namespace). 

The error message right now is ambiguous - the user has no idea what privileges he needs - he'll have to hunt that down in the docs. Granted, just getting the minimum privileges here ("get at least one namespace") won't be enough to really use Kiali, but maybe this error message should help the user a little more. Maybe tell them "Look at the Kiali documentation for the proper permissions needed to access Kiali, including the requirement that you must be able to get at least one namespace". I can't think of a good way to say it - but I know the way it is right now should change :)

Reply:
What about this one?

Reply:
We could say "Not enough privileges to login. Check Kiali's RBAC documentation", hoping users will visit https://kiali.io/documentation/latest/configuration/rbac/.